freedomofpress/securedrop
# Eliminate Flask-SQLAlchemy deprecation warnings

## Status

Ready for review

## Description of Changes

Our Python 3 test output is currently littered with pages upon pages of SQLAlchemy deprecation warnings, making it difficult to get to test failures. We won't want these warnings cluttering production logs, either.

This PR eliminates the deprecation warnings with the following changes:

- Replace `Binary` with `LargeBinary` in `securedrop/models.py`. This eliminates `SADeprecationWarning: The Binary class is deprecated and will be removed in a future relase. Please use LargeBinary.`

- Update Flask-SQLAlchemy to 2.4.0 from 2.3.2, eliminating the following warnings:

  - `SADeprecationWarning: The create_engine.convert_unicode parameter  and corresponding dialect-level parameters are deprecated, and will  be removed in a future release.  Modern DBAPIs support Python Unicode natively and this parameter is unnecessary.`

  - `SADeprecationWarning: Use .persist_selectable`

## Testing

- Run `make dev`. Log in as `journalist`, then verify that you can change the account password and log back in. (The `Binary` columns were `Journalist.pw_hash` and `Journalist.pw_salt`.)
- Run `make test`.

Compare the Flask-SQLAlchemy packages (I used `pkgdiff`):
```
5971b9852b5888655f11db634e87725a9031e170f37c0ce7851cf83497f56e53  Flask-SQLAlchemy-2.3.2.tar.gz
0c9609b0d72871c540a7945ea559c8fdf5455192d2db67219509aed680a3d45a  Flask-SQLAlchemy-2.4.0.tar.gz
```

- [The 2.4.0 changelog](https://flask-sqlalchemy.palletsprojects.com/en/2.x/changelog/#version-2-4-0)
- [2.4.0 on PyPI](https://pypi.org/project/Flask-SQLAlchemy/2.4.0/#files)
- [2.3.2 on PyPI](https://pypi.org/project/Flask-SQLAlchemy/2.3.2/#files)

## Deployment

I don't believe this requires any special handling. The `Binary` class has been the same as `LargeBinary` for a long time, and Alembic generates an empty migration for this change. The other warnings involve engine configuration happening behind the scenes, not in our code.

## Checklist

### If you made changes to the server application code:


### If you made non-trivial code changes:
